inspect/rpc/rpc.go

```diff
@@ -0,0 +1,126 @@
+package rpc
+
+import (
+	"context"
+	"net/http"
+	"time"
+
+	"github.com/rs/cors"
+	"github.com/tendermint/tendermint/config"
+	"github.com/tendermint/tendermint/libs/log"
+	tmpubsub "github.com/tendermint/tendermint/libs/pubsub"
+	rpccore "github.com/tendermint/tendermint/rpc/core"
+	rpcserver "github.com/tendermint/tendermint/rpc/jsonrpc/server"
+	"github.com/tendermint/tendermint/state"
+	"github.com/tendermint/tendermint/state/indexer"
+	"github.com/tendermint/tendermint/types"
+)
+
+// Server defines parameters for running an Inspect rpc server.
+type Server struct {
+	Addr    string // TCP address to listen on, ":http" if empty
+	Handler http.Handler
+	Logger  log.Logger
+	Config  *config.RPCConfig
+}
+
+// Routes returns the set of routes used by the Inspect server.
+func Routes(store state.Store, blockStore state.BlockStore, eventSinks []indexer.EventSink) rpccore.RoutesMap {
+	env := &rpccore.Environment{
+		EventSinks: eventSinks,
+		StateStore: store,
+		BlockStore: blockStore,
+	}
+	return rpccore.RoutesMap{
+		"blockchain":       rpcserver.NewRPCFunc(env.BlockchainInfo, "minHeight,maxHeight", true),
+		"consensus_params": rpcserver.NewRPCFunc(env.ConsensusParams, "height", true),
+		"block":            rpcserver.NewRPCFunc(env.Block, "height", true),
+		"block_by_hash":    rpcserver.NewRPCFunc(env.BlockByHash, "hash", true),
+		"block_results":    rpcserver.NewRPCFunc(env.BlockResults, "height", true),
+		"commit":           rpcserver.NewRPCFunc(env.Commit, "height", true),
+		"validators":       rpcserver.NewRPCFunc(env.Validators, "height,page,per_page", true),
+		"tx":               rpcserver.NewRPCFunc(env.Tx, "hash,prove", true),
+		"tx_search":        rpcserver.NewRPCFunc(env.TxSearch, "query,prove,page,per_page,order_by", false),
+		"block_search":     rpcserver.NewRPCFunc(env.BlockSearch, "query,page,per_page,order_by", false),
+	}
```

Comment:
we should probably try and construct this in the rpc/core package so that we don't end up with different routes in inspect mode and in the rpc pacakge?

Reply:
Good call.

Reply:
Looking at this again, this will require exposing a number of new functions out of the RPC package. I began to do this but it changes a lot of the `env` API. We get a good bit of code reuse here just by using the `env.*` methods already. While we continue to figure out a better direction for the RPC, I'd prefer to keep the mild code duplication to prevent quite so much code churn.

Reply:
Let's make a ticket to clean this up, so we don't loose track of it.

Reply:
#6855